apps/studio/src/server/modules/resource/resource.router.ts

```diff
@@ -146,21 +186,82 @@ export const resourceRouter = router({
   move: protectedProcedure
     .input(moveSchema)
     .mutation(async ({ input: { movedResourceId, destinationResourceId } }) => {
+      const isValid = await validateUserPermissions({
+        from: movedResourceId,
+        to: destinationResourceId,
+        userId: ctx.user.id,
+        siteId,
+      })
+
+      if (!isValid) {
+        throw new TRPCError({
+          code: "FORBIDDEN",
+          message:
+            "Please ensure that you have the required permissions to perform a move!",
+        })
+      }
+
+      return await db.transaction().execute(async (tx) => {
+        const toMove = await tx
+          .selectFrom("Resource")
+          .where("id", "=", movedResourceId)
+          .select(["id", "siteId"])
+          .executeTakeFirst()
+        if (!toMove) {
+          throw new TRPCError({ code: "BAD_REQUEST" })
+        }
+        const parent = await tx
+          .selectFrom("Resource")
+          .where("id", "=", destinationResourceId)
+          .select(["id", "type", "siteId"])
+          .executeTakeFirst()
+
+        if (!parent || parent.type !== "Folder") {
+          throw new TRPCError({ code: "BAD_REQUEST" })
+        }
+
+        if (movedResourceId === destinationResourceId) {
+          throw new TRPCError({ code: "BAD_REQUEST" })
+        }
+
+        if (toMove.siteId !== parent.siteId) {
+          throw new TRPCError({ code: "FORBIDDEN" })
+        }
+
+        await tx
+          .updateTable("Resource")
+          .where("id", "=", String(movedResourceId))
+          .where("Resource.type", "in", ["Page", "Folder"])
+          .set({ parentId: String(destinationResourceId) })
+          .execute()
+        return tx
+          .selectFrom("Resource")
+          .where("id", "=", String(movedResourceId))
+          .select([
+            "Resource.parentId",
+            "Resource.id",
+            "Resource.type",
+            "Resource.permalink",
+            "Resource.title",
+          ])
+          .executeTakeFirst()
+      })
+
+      if (!isValid) {
+        throw new TRPCError({
+          code: "FORBIDDEN",
+          message:
+            "Please ensure that you have the required permissions to perform a move!",
+        })
+      }
```

Comment:
nit: can't really tell from the code review interface but this (and below) appears to be unreachable code?